yb-voyager/src/tgtdb/yugabytedb.go

```diff
@@ -514,13 +582,100 @@ func (yb *TargetYugabyteDB) importBatch(conn *pgx.Conn, batch Batch, args *Impor
 		return res.RowsAffected(), err
 	}
 
-	err = yb.recordEntryInDB(tx, batch, res.RowsAffected())
+	// 5. Record the import in the DB.
+	err = yb.recordEntryInDB(conn, batch, res.RowsAffected())
 	if err != nil {
 		err = fmt.Errorf("record entry in DB for batch %q: %w", batch.GetFilePath(), err)
 	}
 	return res.RowsAffected(), err
 }
 
+// importBatchFastRecover is used to import a batch which was previously tried via fast path but failed
+func (yb *TargetYugabyteDB) importBatchFastRecover(conn *pgx.Conn, batch Batch, args *ImportBatchArgs) (int64, error) {
+	// 1. Check if the split is already imported.
+	alreadyImported, rowsAffected, err := yb.isBatchAlreadyImported(conn, batch)
+	if err != nil {
+		return 0, err
+	}
+	if alreadyImported {
+		return rowsAffected, nil
+	}
+
+	// 2. Open the batch file as datafile
+	df, err := batch.OpenAsDataFile()
+	if err != nil {
+		return 0, fmt.Errorf("open file %s: %w", batch.GetFilePath(), err)
+	}
+	defer df.Close()
+
+	// 3. Read the batch file line by line(row) and build INSERT statment for it based on PK conflict action
+	var rowsIgnored int64 = 0
+	rowsAffected = 0
+
+	// [TODO: Optional] Use PreparedStatement Caching on coonPool for better performance
+	insertStmt := args.GetInsertPreparedStmtForBatchImport()
+	for {
+		line, _, readLinErr := df.NextLine()
+		if readLinErr != nil && readLinErr != io.EOF {
+			return 0, fmt.Errorf("read line from file %s: %w", batch.GetFilePath(), err)
+		}
+
+		/*
+			Both the cases are handled:
+			1. line="" + EOF error
+			2. line!=""(last line) + EOF error
+		*/
+		if line == "" { // handles case 1
+			if readLinErr == io.EOF {
+				break
+			} else {
+				// skipping if any empty line (not expected from batch file)
+				continue
+			}
+		}
+
+		fields := strings.Split(line, args.Delimiter)
+		stmtArgs := make([]interface{}, len(fields))
+		for i, val := range fields {
+			stmtArgs[i] = val
+		}
+
+		// 5. Execute the INSERT statement one by one
+		res, err := conn.Exec(context.Background(), insertStmt, stmtArgs...)
+		if err != nil {
+			return 0, fmt.Errorf("insert into target table: %w", err)
+		}
+
+		if res.RowsAffected() > 0 {
+			rowsAffected++
+		} else {
+			rowsIgnored++
+		}
+		if readLinErr == io.EOF { // handles case 2
+			log.Infof("reached end of file %s", batch.GetFilePath())
+			break
+		}
+	}
+	log.Infof("total rows inserted for batch %q: %d", batch.GetFilePath(), rowsAffected)
```

Comment:
Can we do as assertion here - rowsAffected +rowsIgnored should be equal to batch.RecordCount. 
If not, we should error out.  

Maybe I'm being a bit paranoid here, what do you think?

Reply:
yes lets not add it here. We can add it on top layer where we call import batch irrespective of path.